spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/pubsub/it/PubSubTemplateIntegrationTests.java

```diff
@@ -110,9 +109,18 @@ public void testPullAndAck() {
 
 			PubSubTemplate pubSubTemplate = context.getBean(PubSubTemplate.class);
 
-			pubSubTemplate.publish(topicName, "free-hand");
-			pubSubTemplate.publish(topicName, "valedictory");
-			pubSubTemplate.publish(topicName, "the-runaway");
+			List<Future<String>> futures = new ArrayList<>();
+			futures.add(pubSubTemplate.publish(topicName, "free-hand"));
+			futures.add(pubSubTemplate.publish(topicName, "valedictory"));
+			futures.add(pubSubTemplate.publish(topicName, "the-runaway"));
+
+			futures.parallelStream().forEach( f -> {
+				try {
+					f.get();
```

Comment:
You can't do this `get()` which is infinite by nature. You will block CI server forever if there is something wrong with the target Pub/Sub system.

I wonder if there is some callback where we can use a `CountDownLatch` with its `await(10_000)` afterwards to assert that we really get a result during reasonable timeout.

Reply:
There is `Future.get(long timeout, TimeUnit unit)` which allows you to give that time limit, too.

Reply:
Well, looking to the test-case closely I don't think that we need such a kung-fu about the publish `Future`.

Our test relies on the `PubSubTemplate.pull()`. Seems for me it must be blocked at least for some time. Isn't it?
Or what am I missing?

Thanks

Reply:
I'm going to change this test significantly. In it's current implementation it doesn't actually test what we want it to test. For example, there is no only-once delivery guarantee for messages, on which this test relies when it tries to count how many messages we received.

Reply:
I don't really see the need for this test to be changed "significantly".
In practice, we find that the messages do end up being delivered once.

Reply:
I think that's one of the reasons this test is failing from time to time.

Reply:
So we get more than 5 messages?
Do you have a sample Travis output?

Reply:
@joaoandremartins I've seen it locally